File: 12_String_Hashing/dvaput.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1e9 + 7;
const int BAS = 200;

long long ok(const string& s, int n) {
    if (n == 0) return true;
    if (n == 1) {
        set<char> uniq(s.begin(), s.end());
        return uniq.size() != s.size();
    }
    
    long long cur = 0;
    long long rem = 1;
    for (int i = 0; i < n - 1; i++) {
        rem = (rem * BAS) % MOD;
    }

    unordered_map<long long, bool> seen;
    for (int i = 0; i < n; i++) {
        cur = (cur * BAS) % MOD;
        cur = (cur + s[i]) % MOD;
    }
    
    seen[cur] = true;
    for (int i = n; i < (int)s.size(); i++) {
        cur = (cur - s[i - n] * rem) % MOD;
        if (cur < 0) cur += MOD;
        
        cur = (cur * BAS) % MOD;
        cur = (cur + s[i]) % MOD;

        if (seen.count(cur)) return true;
        seen[cur] = true;
    }

    return false;
}

int high_bound(const string& a) {
    int l = 0, r = a.size();
    int ans = -1;
    while (l < r) {
        int mid = (l + r) / 2;
        if (ok(a, mid)) {
            ans = mid;
            l = mid + 1;
        } else {
            r = mid;
        }
    }
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    
    int h;
    cin >> h;
    cin.ignore();
    
    string s;
    getline(cin, s);

    cout << high_bound(s) << "\n";

    return 0;
}

File: 12_String_Hashing/dvaput.py
import sys
from collections import *
sys.setrecursionlimit(10**5)
itr = (line for line in sys.stdin.read().strip().split('\n'))
INP = lambda: next(itr)
def ni(): return int(INP())
def nl(): return [int(_) for _ in INP().split()]


MOD = 10**16 + 61
BAS = 200



def ok(s, n):
    if n == 0:
        return True
    if n == 1:
        if len(set(list(s))) == len(list(s)):
            return False
        else:
            return True
    cur = 0
    rem = pow(BAS, n-1, MOD)
    seen = {}
    for i in range(n):
        #print(i, s[i])
        cur = (cur * BAS) % MOD
        cur = (cur + ord(s[i])) % MOD
    
    #print(s[:n])
    seen[cur] = True
    #print(cur)
    for i in range(n, len(s) - 1):
        #remove
        cur = (cur - ord(s[i - n]) * (rem)) % MOD
        cur = (cur * BAS + ord(s[i])) % MOD
        
        #print(i, i-n, s[i-n: i], s[i-n], s[i], cur)
        if cur in seen:
            return True
        seen[cur] = True

    return False

#Binary search for higher bound
def high_bound(a):
    l, r,  = 0, len(a)
    ans = -1
    while l < r:
        mid = (l + r) // 2
        #print(l, r, mid, a)
        if ok(a, mid):
            ans = mid
            l = mid+1
        else:
            r = mid
    return ans
        
       
h = ni() 
s = INP()
l = [ord(s[i]) for i in range(len(s))]

print(high_bound(s))